### ISSUE TITLE
IO#read_nonblock throwing EOFError with "exception: false"

### ISSUE THREAD

**Issue by HoneyryderChuck** [jruby/jruby#4827]
I'm using exceptionless non-blocking IO API with Jruby 9.1.13.0. In a busy loop I have, it returns (predictably) `:wait_readable`, until it doesn't, and I get the message: 

```
 EOFError: No message available
```

I thought that, by using `exception: false`, that the nonblocking APIs can't throw exception, but rather, `:wait_readable/writable` when buffer is unavailable, and `nil` when IO is closed (equivalent to `EOFError` in older rubies). In fact, this behaviour is consistent in ruby 2.3.0 and newer. 

However, I can't seem to find documentation validating this. As of ruby-doc, this is only enforced since 2.4.2, and only for `#read_nonblock`. But I think this is more a case of the docs not being up-to-date.

**Comment by HoneyryderChuck** [comment]
Also, dunno if related, but although `EOF`ed, it's still selectable:

```ruby
loop do
  r, w = IO.select([r], [w])
  r.each do |reader|
    begin
      read.read_nonb...
    rescue EOFError
    end
  end
end
```

which is another difference with CRuby.

**Comment by headius** [comment]
I believe you are correct in your expectations. I'll have a look.

**Comment by headius** [comment]
If you are able to reproduce it, pass -Xbacktrace.style=full to JRuby and post the resulting trace. It would help me see where it's going wrong.

**Comment by headius** [comment]
I believe I have fixed this in fcd7d7357a but I'd like to add or find a test that can trigger this behavior. Any additional information you can provide would be helpful!

**Comment by HoneyryderChuck** [comment]
I believe that the backtrace you need is this:

```
 java/lang/Thread.java:1552:in `getStackTrace'
org/jruby/runtime/backtrace/TraceType.java:246:in `getBacktraceData'
org/jruby/runtime/backtrace/TraceType.java:47:in `getBacktrace'
org/jruby/RubyException.java:235:in `prepareBacktrace'
org/jruby/exceptions/RaiseException.java:216:in `preRaise'
org/jruby/exceptions/RaiseException.java:183:in `preRaise'
org/jruby/exceptions/RaiseException.java:111:in `<init>'
org/jruby/Ruby.java:4138:in `newRaiseException'
org/jruby/Ruby.java:4093:in `newEOFError'
org/jruby/RubyIO.java:2907:in `getPartial'
org/jruby/RubyIO.java:2812:in `doReadNonblock'
org/jruby/RubyIO.java:2803:in `read_nonblock'
org/jruby/RubyIO$INVOKER$i$0$2$read_nonblock.gen:-1:in `call'
org/jruby/internal/runtime/methods/JavaMethod.java:729:in `call'
org/jruby/internal/runtime/methods/DynamicMethod.java:216:in `call'
org/jruby/runtime/callsite/CachingCallSite.java:225:in `call'
```

To reproduce this in a script is a bit non-trivial, as this doesn't happen all the time in my tests (just in a particular one, where more than the usual payload passed between sockets). Maybe I can test this again once `9.1.14.0` is out the door.
